(Towards #1338) Implementation of an OMPTaskwaitTrans

LonelyCat124 opened the pull request:
Initial implementation of the OMPTaskwaitTrans is implemented here.

I've fixed all the pycodestyle and pylint issues I've found other than these, which I'm not clear if I can resolve or not:
```
transformations.py:677:4: R0914: Too many local variables (19/15) (too-many-locals)
transformations.py:677:4: R0912: Too many branches (15/12) (too-many-branches)
```

I've not yet implemented any tests, nor the `{'fail_on_no_taskloops':True/False}` option that Sergi suggested.

Some of the tests are easy to implement, but others I could do with suggestions of a starting case (from a code perspective).
I need to know how to write a piece of general code with (2+, not sure exactly) loops that I can use as the basis for tests, what do I need to put together for PSyclone to handle this? @sergisiso @arporter @rupertford any ideas on what I need to put together for that?

arporter commented:
The easiest thing to do is write what you want in Fortran (as a string) and then use the `fortran_reader` test fixture to give you the PSyIR:

    psyir = fortran_reader.psyir_from_source(my_string)

arporter commented:
The Fortran will need to be a valid Subroutine (or Program/Function).

LonelyCat124 commented:
Added some first few tests, added the `{'fail_on_no_taskloops':True/False}` option, and deleted a line from `forward_dependence` in `nodes.py` which will definitely cause no issues at all...

LonelyCat124 commented:
Concerns me a little that tests pass xD. Will finish a draft of this off tomorrow am

LonelyCat124 commented:
I'm going to tentatively mark this as ready to review, though I will also leave it as draft for now.

The main thing that might change is the use of `forward_dependence` may change to use @hiker 's new dependence analysis system if I can work out an appropriate way to implement a "forward_dependence" style function for it.

I think I've tested simplified versions of all possible dependency structures for taskloops that I can think of for now (except for non-parallelised loops I guess), so if anyone wants to try to break it as well it would be good to have more tests there.

LonelyCat124 commented:
Removing ready to review tag until the new version is pushed!

LonelyCat124 commented:
New version is completed. If the review wants I can try to resolve the linter issues:
```
src/psyclone/transformations.py:692:4: R0912: Too many branches (17/12) (too-many-branches)
src/psyclone/transformations.py:797:4: R0914: Too many local variables (19/15) (too-many-locals)
src/psyclone/transformations.py:797:4: R0912: Too many branches (18/12) (too-many-branches)
```
by refactoring after review but I'm not sure if it would make the code clearer or not.

LonelyCat124 commented:
From discussion with Tobias it seems like:
```
!$omp parallel
!$omp single
!$omp taskloop nogroup
do i=1, n
  a(i) = 2 * i
end do
!$omp end taskloop
!$omp end single
!$omp do
do i=1,n
  b(i) = a(i) * 3
end do
!$omp end do
!$omp end parallel
```

Is not valid unless I add a `taskwait` at the end of the single. I will implement this now if I can

LonelyCat124 commented:
OpenMP does define: 
> task synchronization construct: A taskwait, taskgroup, or a barrier construct.

However I'm going to assume that implicit barriers do not count for this.

LonelyCat124 commented:
I think there are still some over sensitivities in my `get_forward_dependence`, I will continue to improve it.

LonelyCat124 commented:
I have removed the over sensitivities - i cannot work out why the examples are suddenly failing to build though (they work fine locally)

LonelyCat124 commented:
I think coverage is 100% on this diff (from checking locally with 3.6), so this is ready to review.

LonelyCat124 commented:
Not sure where to ask this, but can I check:
Does the `LoopTrans` class's validate (which i can't find) or the `Loop` class require the Loop to meet OpenMP's canonical loop form (https://www.openmp.org/spec-html/5.0/openmpsu40.html#x63-1260002.9.1)? The main thing I can't think of being covered when I looked last week is that I couldn't see is that I don't think any of them check `incr-expr` (C/C++) or `incr` (Fortran).
I think in theory PSyclone could create invalid code here? @sergisiso what do you think, should we make an issue for this?

sergisiso commented:
LoopTrans is in `src/psyclone/psyir/transformations/loop_trans.py`

I wasn't aware of this restrictions and you are right that this could be validated. Maybe subclassing LoopTrans with OMPLoopTrans and adding this additional restrictions? Which if I read correctly for Fortran are:
- Only BinaryOperation.[ADD, SUB, MUL] + Literals + References are valid in in the incr
- Loop invariant with respect to the outermost associated loop. As discussed the other day I think for a single loop in Fortran, the incr is loop invariant by definition, but I don't think this is true for nested loops then and could be checked in the validate.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        for node in node_list:
+            if node.abs_position <= taskloop.abs_position:
+                continue
+            # Ignore our children
```
s/our children/any descendents of the taskloop directive/

LonelyCat124 replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+            if node.ancestor(OMPParallelDirective) is not parent_parallel:
+                return parent_parallel
+
+            if (isinstance(node, OMPTaskwaitDirective) and
```
Bearing in mind that going up and down the tree can get expensive, please could you cache some of these results (e.g. `taskloop.ancestor(OMPSingleDirective)`) so that we find it just once and then re-use it. It might also help make this `if` a bit shorter :-)

LonelyCat124 replied:
I think I've cached the ones I can, unfortunately the code checks for different types of ancestor a lot :(

LonelyCat124 replied:
I checked `ancestor` docstring and it says it only returns the first instance found. I didn't check the code explicitly so its possible thats wrong.

hiker replied:
No idea what I was thinking here :(  Please ignore ;)

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                node_vars = taskloop_vars
+            elif not isinstance(node, (OMPSerialDirective,
+                                       OMPParallelDirective)):
+                # For all our other node types that cannot be our parents,
```
I don't understand this comment. Perhaps this relates to what the expected relationship between `taskloop` and `root` is? Please could you clarify and maybe add a validation step to check that things are as you expect upon entry to this routine?

LonelyCat124 replied:
Actually it is never a parent of OMPTaskloop by this check:
```
if node.abs_position <= taskloop.abs_position:
                  continue
```

LonelyCat124 replied:
This has been rewritten but hopefully clears up a lot of the confusing behaviour.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                # For all our other node types that cannot be our parents,
+                # they use their own signatures
+                node_vars = VariablesAccessInfo()
+                for child in node.walk(nodes.Node):
```
We do this walk earlier so could we re-use the results from there?

LonelyCat124 replied:
I think we only do this for taskloop previously, not for node.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                                                            (Schedule, Loop)):
+                        node_vars.merge(VariablesAccessInfo(child))
+            else:
+                # If we find node types that could be our parents (are are
```
s/are are/and are/ I think? Please could you also clarify what "known to not be valid options" means?

LonelyCat124 replied:
Removed code.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                access1 = taskloop_vars[sig1]
+                access2 = node_vars[sig1]
+                # If both read only, we can ignore this signature
+                # Otherwise, one of themwrites so return this node as
```
s/themwrites/them writes/

LonelyCat124 replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                    # If we have a different parent serial node, and
+                    # parent_single is not None then our parent_single is our
+                    # dependency, otherwise this node is the dependency
+                    if (taskloop.ancestor(OMPSerialDirective) is not
```
Use cached result for `taskloop.ancestor(OMPSerialDirective)`?

LonelyCat124 replied:
I previously only computed `taskloop.ancestor(OMPSingleDirective)` - the difference here is there is another subclass of `OMPSerialDirective` which is legal to contain these nodes (`OMPMasterDirective`), but does not allow some dependency patterns between regions to be satisfiable.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+    def apply(self, node, options=None):
+        '''
+        Apply an OMPTaskwait Transformation to the supplied node
+        (which must be a OMPParallelDirective). In the generated code this
```
s/be a/be an/

LonelyCat124 replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+          !$OMP END PARALLEL
+
+        :param node: the node to which to apply the transformation.
+        :type node: :py:class:`psyclone.f2pygen.DoGen`
```
Must be a PSyIR node, i.e. `psyclone.psyir.nodes.OMPParallelDirective`

LonelyCat124 replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        :type node: :py:class:`psyclone.f2pygen.DoGen`
+        :param options: a dictionary with options for transformations\
+                        and validation.
+        :type options: dictionary of string:values or None
```
"dict of str:values or None"

LonelyCat124 replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        :type options: dictionary of string:values or None
+        :param bool options["fail_on_no_taskloop"]:
+                indicating whether this should throw an error if no \
+                OMPTaskloop nodes are found in this tree.
```
Please add what the default is.

LonelyCat124 replied:
Using the better rewritten docstring from `validate` now for this

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        # create a memento of the schedule and the proposed transformation
+        keep = Memento(schedule, self, [node])
+
+        # Find all the OpenMP Single & Master regions
```
As this is new code and we've decided to drop Memento, please remove this bit and just have the routine return (None, Node) and add a TODO for #595.

LonelyCat124 replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+
+        self.validate(node, options=options)
+
+        schedule = node.root
```
I'm not sure you want to do this or is it just here for the Memento?

LonelyCat124 replied:
It was for momento, removed.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                        if forward_dep is task_region:
+                            endwaits.append(taskloop)
+                            create_endtaskwait = True
+                        forward_dep = None
```
Could just do `continue` here to make it a bit more obvious?

LonelyCat124 replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+            # dependence_position[this]) then we remove this dependence.
+            # This assumes that taskloop_positions is in ascending order,
+            # which I believe to be true by construction.
+            for i, dep_pos in enumerate(dependence_position):
```
Please could you break out this elimination into a separate routine (could be a private one).

LonelyCat124 replied:
Done as a private static method

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+            # This assumes that taskloop_positions is in ascending order,
+            # which I believe to be true by construction.
+            for i, dep_pos in enumerate(dependence_position):
+                if dep_pos is not None:
```
Perhaps:

    if dep_pos is None:
        continue

so as to reduce the indentation of the rest of the code.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+            # immediately before.
+            for i, dep_pos in enumerate(dependence_position):
+                if dep_pos is not None:
+                    forward_dep = OMPTaskwaitTrans.get_forward_dependence(
```
I'm a bit confused here, don't we already have the forward dependence from earlier?

LonelyCat124 replied:
I didn't cache it before, I will change that now.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                            taskloops[i], node)
+                    fdep_parent = forward_dep.parent
+                    # Find the position of the forward_dep in its parent's
+                    # children list
```
I think you can just do `fdep_parent.children.index(forward_dep)`

LonelyCat124 replied:
Changed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                            unsatisfied = False
+                    create_endtaskwait = create_endtaskwait or unsatisfied
+                if create_endtaskwait:
+                    task_region.children[0].addchild(OMPTaskwaitDirective())
```
Should `children[0]` be `directive_body` or similar?

arporter commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                        if node1.abs_position > taskloop_pos:
+                            unsatisfied = False
+                    create_endtaskwait = create_endtaskwait or unsatisfied
+                if create_endtaskwait:
```
If I've understood this, as soon as we have `unsatisfied = True` then we know we need to create an OMPTaskWaitDirective at the end of this region? That being so I think we could simplify a bit:

    if any(node1.abs_position <= taskloop_pos for node1 in nodes):
        task_region.directive_body.addchild(....)
        break

(but I may be wrong)

LonelyCat124 replied:
Yes, I think thats almost correct, I think its:
```
if all(node1.abs_position <= taskloop_pos for node1 in nodes):
    # do stuff and break
```
as its only if there is no taskwait between this taskloop and the end of the single that we need the extra barrier.

arporter commented on src/psyclone/tests/test_files/gocean1p0/reordered_dependent_invoke.f90:
```diff
@@ -0,0 +1,52 @@
+!-------------------------------------------------------------------------------
```
Please could you use the current license header here.

LonelyCat124 replied:
I think this reordered one is a test I made to try some coverage improvements, but actually I can remove.

arporter commented on src/psyclone/tests/test_files/gocean1p0/reordered_dependent_invoke.f90:
```diff
@@ -0,0 +1,52 @@
+
+PROGRAM dependent_invoke_test
+
+  ! Fake Fortran program for testing aspects of
```
Please extend this to say what aspect this file is testing. (Also, are you sure there isn't one that already does this?)

LonelyCat124 replied:
Removed

arporter commented on src/psyclone/tests/test_files/gocean1p0/reordered_compute_cu_mod.f90:
```diff
@@ -0,0 +1,114 @@
+!> \brief Compute the mass flux in the x direction, cu
```
Please add the license/copyright header and extend the comment to say what this file is testing. You'll also need to remove the comment that says what the code originally looked like as that no longer matches the implementation (presumably because this is now a different kernel that doesn't exist in the original NemoLite2D code?).

LonelyCat124 replied:
Done

LonelyCat124 commented on .github/workflows/python-package.yml:
```diff
@@ -57,7 +57,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [2.7, 3.5, 3.8]
```
Switched it back

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -63,7 +64,9 @@
     Directive, ACCLoopDirective, OMPDoDirective, OMPParallelDoDirective, \
     ACCDataDirective, ACCEnterDataDirective, OMPDirective, \
-    ACCKernelsDirective, OMPTaskloopDirective
+    ACCKernelsDirective, OMPTaskloopDirective, OMPSerialDirective, \
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+    >>> schedule.view()
+    '''
+    def __str__(self):
+        rval = ("Adds 'OpenMP TASKWAIT' directives to a loop to satisfy "
```
I changed this to read correctly. I'm not sure they do, but I couldn't think of a short way to describe this transformation (Adding a single directive to a loop still fully explains what those transformations do)

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+    >>>
+    >>> # Apply the OpenMP Taskloop transformation to *every* loop
+    >>> # in the schedule.
+    >>> # This ignores loop dependencies. These must be manually handled
```
Fixed this.

LonelyCat124 replied:
Also updated this for tasklooptrans

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        :param node: the PSyIR node to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        '''
+        Validity checks for input arguments.
+
+        :param node: the PSyIR node to validate.
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        :type options: dictionary of string:values or None
+        :param bool options["fail_on_no_taskloop"]:
+                indicating whether this should throw an error if no \
+                OMPTaskloop nodes are found in this tree.
```
I've improved this.

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+            fail_on_no_taskloop = options.get("fail_on_no_taskloop", True)
+        # Check the supplied node is an OMPParallelDirective
+        if not isinstance(node, nodes.OMPParallelDirective):
+            raise TransformationError("OMPTaskwaitTrans was supplied a \"{0}\""
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+
+        # Walk the tree to find any OMPTaskloopTrans
+        if fail_on_no_taskloop and node.walk(OMPTaskloopDirective) == []:
+            raise TransformationError("OMPTaskwaitTrans was supplied a "
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                                      "OMPTaskloopDirectives")
+
+        # Check that all of the dependencies are satisfiable
+        taskloops = node.walk(OMPTaskloopDirective)
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        taskloops = node.walk(OMPTaskloopDirective)
+
+        for taskloop in taskloops:
+            forward_dep = OMPTaskwaitTrans.get_forward_dependence(taskloop,
```
I've added one, I could add more detail but I'd just have to describe the get_forward_dependence function if so.

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                                                                  node)
+            if forward_dep is None:
+                continue
+            # If the forward dependence is always in the same Parallel
```
First two lines of this comment block are for old code that I think removed, so I've removed those lines

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                    forward_dep.ancestor(OMPSerialDirective)):
+                # They're not in the same serial region. Check if our
+                # taskloop is in a waiting Single Directive
+                valid = True
```
Yup, fixed!

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                valid = valid and isinstance(ancestor,
+                                             nodes.OMPSingleDirective)
+                if valid:
+                    valid = valid and (not ancestor.nowait)
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                # safely guarantee our forward dependency so throw an
+                # error
+                if not valid:
+                    raise TransformationError("Couldn't satisfy the "
```
I think I've done this now.

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+    @staticmethod
+    def get_forward_dependence(taskloop, root):
+        '''
+        Returns the next forward dependence for a taskloop using the new
```
Changed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        The forward dependency is never a child of taskloop, and must have
+        abs_position > taskloop.abs_position
+
+        :param taskloop: the taskloop node to search for the
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+
+        :param taskloop: the taskloop node to search for the
+                         forward_dependence.
+        :type taskloop: :py:class:`psyclone.psyir.nodes.OMPTaskloopDirective
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                         forward_dependence.
+        :type taskloop: :py:class:`psyclone.psyir.nodes.OMPTaskloopDirective
+        :param root: the tree in which to search for the forward_dependence.
+        :type root: :py:class:`psyclone.f2pygen.DoGen`
```
I don't know - I copied this from the main `get_forward_dependence` function I think. I'l lchange it for `Node` for now.

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        :type root: :py:class:`psyclone.f2pygen.DoGen`
+
+        :returns: the forward_dependence of taskloop.
+        :rtype: :py:class:`psyclone.f2pygen.DoGen`
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                               OMPTaskwaitDirective, OMPSingleDirective,
+                               OMPParallelDirective))
+        taskloop_vars = VariablesAccessInfo()
+        for child in taskloop.walk(nodes.Node):
```
I will add a comment explaining this. Essentially the functionality that I need is not quite provided in `dependency_tools`. The `VariablesAccessInfo` that results from using the dependency_tools directly is the read/write accesses to the loop variable (by the loops), which means that since loop variables are probably reused by each loop, all loops point dependencies to each other. The easiest way I could think of to do this is to implement it myself, where I skip over all Loops explicitly.

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        taskloop_signatures = taskloop_vars.all_signatures
+        # Find our parent serial region if it has a barrier
+        parent_single = taskloop.ancestor(OMPSingleDirective)
+        if parent_single is not None and parent_single.nowait:
```
Fixed

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        if parent_single is not None and parent_single.nowait:
+            parent_single = None
+        # Find our parent parallel region
+        parent_parallel = taskloop.ancestor(OMPParallelDirective)
```
Done

LonelyCat124 commented:
Ok, I have resolved most comments (one refactor to do) and fixed the tests. I expect that some of the coverage will have dropped so I will look into that next.

I have some linter issues I'm not clear how to resolve:
```
src/psyclone/transformations.py:691:60: W0640: Cell variable fwriter defined in loop (cell-var-from-loop)
src/psyclone/transformations.py:691:68: W0640: Cell variable taskloop defined in loop (cell-var-from-loop)
src/psyclone/transformations.py:693:60: W0640: Cell variable fwriter defined in loop (cell-var-from-loop)
src/psyclone/transformations.py:694:64: W0640: Cell variable forward_dep defined in loop (cell-var-from-loop)
```
where I'm using some variables in a lambda (which is being passed to `LazyString`. Are these save to ignore or is there some way to fix?

arporter commented:
re. the linting errors: I haven't actually used LazyString yet so I defer to @sergisiso and @rupertford  who have...

sergisiso commented:
> I have some linter issues I'm not clear how to resolve

I think what it says is to be careful with the lambda capturing variables since these are references and are going to reference to somewhere else in the next iteration of the loop. But in this case this is not a problem since this is a "raise" so we know there won't be a next iteration. So I would say it is safe to disable the warning with this explanation or alternatively break the loop at this point and raising the exception after the loop if it has not succeeded may also get rid of the pylint warning.

hiker commented on src/psyclone/transformations.py:
```diff
@@ -563,6 +566,422 @@ def apply(self, node, options=None):
+                for child in node.walk(nodes.Node):
+                    if child is not node and not isinstance(child,
+                                                            (Schedule, Loop)):
+                        node_vars.merge(VariablesAccessInfo(child))
```
Just a quick comment: creating a new instance of VariablesAccessInfo is not necessary, you can simply do:
```
   child.reference_accesses(node_vars)
````
This will add the new accesses to the existing variable (and saves the overhead of creating a new instance, and then merging them together).

LonelyCat124 replied:
Fixed, thanks!

LonelyCat124 commented:
Should be ready for review again now :)

LonelyCat124 commented:
Hmm, not sure why I can't comment on the new review comments but I'll try to go through them today. Probably I'll have to put my responses in this thread @arporter 

> Please add a comment, e.g. "Ensure loop dependencies are satisfied." More generally, how can we prevent a user from generating incorrect code? We could check dependencies at code-generation time but that won't work if there are places where we need to manually override such checks. Perhaps we should have a meta-transformation that performs these various steps and that's the one we publicise to users?

I don't want to have any restrictions on what is applied to loops in a region (as it may well be only some loops want to use tasks). I think I could add a `_dependency_checked` member to taskloops, and require that to be set to True during its validation. I can't really make the access to it private though, so there's nothing to stop a user just setting it to be True themselves, but the idea would be that a region containing a taskloop must have the taskwaitTrans applied before code generation.
